File: src/java/org/voegtle/weatherstation/server/logic/data/RepairJob.java
package org.voegtle.weatherstation.server.logic.data;

import org.voegtle.weatherstation.server.persistence.SmoothedWeatherDataSet;

import java.util.ArrayList;
import java.util.List;

public class RepairJob {

  public class RepairStep {
    public double temperature = 0.0;
    public double humidity = 0.0;
    public double rain = 0.0;

    public RepairStep() {
    }
  }

  private SmoothedWeatherDataSet first;
  private SmoothedWeatherDataSet last;
  private final List<SmoothedWeatherDataSet> defectDataSets = new ArrayList<>();
  private final RepairStep step = new RepairStep();

  public RepairJob() {
  }

  public SmoothedWeatherDataSet getFirst() {
    return first;
  }

  public void setFirst(SmoothedWeatherDataSet first) {
    this.first = first;
  }

  public void setLast(SmoothedWeatherDataSet last) {
    this.last = last;
  }

  public List<SmoothedWeatherDataSet> getDefectDataSets() {
    return defectDataSets;
  }

  public void addDefectDataSet(SmoothedWeatherDataSet dataset) {
    defectDataSets.add(dataset);
  }

  public boolean containsData() {
    return defectDataSets.size() > 0;
  }

  public void calculateStep() {
    if (first == null || last == null) {
      return;
    }
    if (defectDataSets.size() == 0) {
      return;
    }

    step.humidity = (last.getOutsideHumidity() - first.getOutsideHumidity()) / (defectDataSets.size() + 1);
    step.temperature = (last.getOutsideTemperature() - first.getOutsideTemperature()) / (defectDataSets.size() + 1);
    step.rain = (last.getRainCounter() - first.getRainCounter()) / (defectDataSets.size() + 1);
  }

  public RepairStep getStep() {
    return step;
  }

}

File: src/main/java/org/voegtle/weatherstation/server/logic/data/RepairJob.kt
package org.voegtle.weatherstation.server.logic.data

import org.voegtle.weatherstation.server.persistence.entities.SmoothedWeatherDataSet

import java.util.ArrayList

class RepairJob {

  var first: SmoothedWeatherDataSet? = null
  var last: SmoothedWeatherDataSet? = null
  val defectDataSets = ArrayList<SmoothedWeatherDataSet>()
  val step = RepairStep()

  inner class RepairStep {
    var temperature = 0.0
    var humidity = 0.0
    var rain = 0.0
    var insideTemperature: Double? = null
    var insideHumidity: Double? = null
    var kwh: Double? = null
  }

  fun addDefectDataSet(dataset: SmoothedWeatherDataSet) {
    defectDataSets.add(dataset)
  }

  fun containsData(): Boolean {
    return defectDataSets.size > 0
  }

  fun calculateStep() {
    if (first == null || last == null) {
      return
    }
    if (defectDataSets.size == 0) {
      return
    }

    step.humidity = ((last!!.outsideHumidity!! - first!!.outsideHumidity!!) / (defectDataSets.size + 1)).toDouble()
    step.temperature = ((last!!.outsideTemperature!! - first!!.outsideTemperature!!) / (defectDataSets.size + 1)).toDouble()
    if (last!!.rainCounter == null) {
      step.rain = 0.0
    } else {
      step.rain = ((last!!.rainCounter!! - first!!.rainCounter!!) / (defectDataSets.size + 1)).toDouble()
    }

    step.insideHumidity = spreadEqually(first!!.insideHumidity, last!!.insideHumidity, defectDataSets.size)
    step.insideTemperature = spreadEqually(first!!.insideTemperature, last!!.insideTemperature, defectDataSets.size)
    step.kwh = spreadEqually(first!!.kwh, last!!.kwh, defectDataSets.size)
  }

  private fun spreadEqually(firstValue: Double?, lastValue: Double?, numberOfSets: Int): Double? {
    return if (firstValue == null || lastValue == null) {
      null
    } else (lastValue - firstValue) / (numberOfSets + 1)
  }

  private fun spreadEqually(firstValue: Float?, lastValue: Float?, numberOfSets: Int): Double? {
    return if (firstValue == null || lastValue == null) {
      null
    } else ((lastValue - firstValue) / (numberOfSets + 1)).toDouble()
  }

}
